external-crates/move/crates/move-package/resolve_pkg_name_nested_dep_conflict/Move.resolved

```diff
@@ -0,0 +1,5 @@
+When resolving dependencies for package Root, conflicting dependencies found:
+At B -> A
+	ADep-resolved = { local = "deps_only/ADep", addr_subst = { "A" = "00000000000000000000000000000007" } }
+At C -> A
+	ADep = { local = "deps_only/ADep", addr_subst = { "A" = "00000000000000000000000000000042" } }
```

Comment:
In this test, the address substitutions don't match, but also the names don't match -- are we expecting the names to match? (It seems like we might be testing something slightly different if they don't).

Reply:
This one is based on [diamond_problem_dep_nested_dep_conflict](https://github.com/MystenLabs/sui/blob/308dad00c432ef76d6fe3d9183acb65e76c88433/external-crates/move/crates/move-package/tests/test_sources/diamond_problem_dep_nested_dep_conflict/Move.resolved) and is intended to check for address substitution (but actually checks both and will fail with the same message).

But good catch! The reason names are different is because package `A` with dependency `ADep` is resolved externally in the other branch, which means its subgraph is fully reconstructed from the lockfile so name resolution hook doesn't kick in. This actually highlights the issue (2) I described in https://github.com/MystenLabs/sui/pull/14178 and why we will need to force lockfile regeneration when on-chain hooks get enabled.

I've improved this test by adding a separate external resolver script for this https://github.com/MystenLabs/sui/pull/15315/commits/9c9781a08e419ef6c6dc11c1c251d7d52cc91f2a. 

Also while looking into this I've noticed a separate problem -- the `deps_equal` function which checks all dependencies of two packages match didn't respect the custom `PartialEq` implementation for `Package` https://github.com/MystenLabs/sui/blob/5e2e4985ebf49e5e3723582d4035e29bb4eb66ef/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L124-L132 and the above test would fail even with correct address substitution because the resolver is set on one and not the other. I've fixed this with https://github.com/MystenLabs/sui/pull/15315/commits/5e2e4985ebf49e5e3723582d4035e29bb4eb66ef